File: PROGRAMMING/PYTHON CODE/mercurysensor.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt, medfilt
# 1. Simulate Mercury (Hg²⁺) sensor data (ppb concentration)
np.random.seed(42)
time = np.linspace(0, 60, 300)  # 60 seconds, 300 samples
true_mercury_ppb = 1.0 + 0.5 * np.sin(0.3 * np.pi * time / 10)  # base clean signal
noise = np.random.normal(0, 0.0005, size=time.shape)  # random noise
spikes = np.zeros_like(time)
spikes[np.random.randint(0, len(time), 10)] = np.random.uniform(0.001, 0.002, 10)  # occasional noise spikes
raw_signal = true_mercury_ppb + noise + spikes
# 2. Moving Average Filter
def moving_average(data, window_size=5):
    return np.convolve(data, np.ones(window_size) / window_size, mode='same')
filtered_ma = moving_average(raw_signal, window_size=5)
# 3. Median Filter
filtered_med = medfilt(raw_signal, kernel_size=5)
# 4. Butterworth Low-Pass Filter
def butter_lowpass_filter(data, cutoff, fs, order=5):
    nyq = 0.5 * fs  # Nyquist Frequency
    normal_cutoff = cutoff / nyq
    b, a = butter(order, normal_cutoff, btype='low', analog=False)
    y = filtfilt(b, a, data)
    return y
fs = 5.0  # Sampling frequency (Hz)
cutoff = 0.5  # Cutoff frequency (Hz)
filtered_butter = butter_lowpass_filter(raw_signal, cutoff, fs, order=4)
# 5. Visualization
plt.figure(figsize=(12, 8))
# Unfiltered Data
plt.subplot(2, 1, 1)
plt.plot(time, raw_signal, color='gray', alpha=0.7, label='Unfiltered (Raw Data)')
plt.plot(time, true_mercury_ppb, color='blue', linewidth=2, label='True Mercury Signal')
plt.title("Mercury (Hg²⁺) Sensor Data - Water Treatment (Unfiltered)")
plt.xlabel("Time (s)")
plt.ylabel("Concentration (ppb)")
plt.legend()
plt.grid(True)
# Filtered Data Comparison
plt.subplot(2, 1, 2)
plt.plot(time, raw_signal, color='gray', alpha=0.3, label='Raw Data')
plt.plot(time, filtered_ma, 'r-', linewidth=1.5, label='Moving Average')
plt.plot(time, filtered_med, 'g-', linewidth=1.5, label='Median Filter')
plt.plot(time, filtered_butter, 'b-', linewidth=2, label='Butterworth (filtfilt)')
plt.title("Filtered Mercury (Hg²⁺) Data - Comparison of Techniques")
plt.xlabel("Time (s)")
plt.ylabel("Concentration (ppb)")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

File: PROGRAMMING/MATLAB CODE/mercurysensor.m
clc;
clear;
close all;
%1.Simulate Lead sensor data    
fs=10;
t=0:1/fs:100;
true_mercury=2+0.3*sin(0.1*t)+0.1*randn(size(t));
%Add random sensor noise(simulating measurement error)  
noise=0.2*randn(size(t));
mercury_noisy=true_mercury+noise;
%2.Moving Average Filter
window_size=5; %Number of samples in windows
b=(1/window_size)*ones(1,window_size);
a=1;
mercury_moving_avg=filter(b,a,mercury_noisy);
%3.Median Filter
windowSize_med=5;
mercury_median=medfilt1(mercury_noisy,windowSize_med);
%4.Butterworth Low-pass Filter
cutoff_freq=0.1; %Normalized cutoff frequency
[b_butter,a_butter]=butter(4,cutoff_freq);
mercury_butter=filter(b_butter,a_butter,mercury_noisy);
%5.Plot results
figure('Name','Mercury Sensor Data Filtering','NumberTitle','off');
subplot(4,1,1);
plot(t,mercury_noisy,'r');
title('Unfiltered Mercury Sensor Data(Noisy)');
xlabel('Time (s)');
ylabel('Mercury Concentration (ppb)');
grid on;    
subplot(4,1,2);
plot(t,mercury_moving_avg,'b');
title('Mercury Sensor Data - Moving Average Filter');
xlabel('Time (s)');
ylabel('Mercury Concentration (ppb)');
grid on;
subplot(4,1,3);
plot(t,mercury_median,'g');
title('Filtered with Median filter');
xlabel('Time (s)');
ylabel('Mercury Concentration (ppb)');
grid on;
subplot(4,1,4);
plot(t,mercury_butter,'m');
title('Filtered with Butterworth Low-Pass Filter (filtfilt)');
xlabel('Time (s)');
ylabel('Mercury Concentration (ppb)');
grid on;